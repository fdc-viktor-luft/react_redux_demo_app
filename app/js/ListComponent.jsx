import React from 'react'
import {List} from 'immutable';
import {connect} from 'react-redux';

import {ListGroup} from 'react-bootstrap';
import {ListGroupItem} from 'react-bootstrap';

const DisconnectedListComponent = React.createClass({
  propTypes: {
    list: React.PropTypes.instanceOf(List).isRequired
  },

  render() {
    return (
      <ListGroup fill>
        {this.props.list.map((value, index) =>
            <ListGroupItem key={index}>
              {value}
            </ListGroupItem>
        ).toArray()}
      </ListGroup>
    );
  }

});

export const ListComponent = connect(function(state: any) {
    return {
        list: state.list
    };
})(DisconnectedListComponent);
