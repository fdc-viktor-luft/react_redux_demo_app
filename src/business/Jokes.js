// @flow
import { List } from 'immutable';

export class Jokes {
    _listOfJokes: List<string>;

    constructor(initialListOfJokes: List<string> = new List()) {
        this._listOfJokes = initialListOfJokes;
    }

    addNewJoke(joke: string): Jokes {
        return new Jokes(this._listOfJokes.push(joke));
    }

    removeJokeAtPosition(index: number): Jokes {
        return new Jokes(this._listOfJokes.delete(index));
    }

    getJokes(): List<string> {
        return this._listOfJokes;
    }
}
