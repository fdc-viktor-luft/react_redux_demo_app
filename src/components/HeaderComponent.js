// @flow
import React from 'react';
import { PageHeader } from 'react-bootstrap';

export class HeaderComponent extends React.Component {
    render() {
        return <PageHeader>React + Redux Demo</PageHeader>;
    }
}
