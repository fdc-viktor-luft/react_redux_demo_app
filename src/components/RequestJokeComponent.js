// @flow
import React from 'react';
import { connect } from 'react-redux';
import { Panel, FormGroup,  Button, ControlLabel, FormControl } from 'react-bootstrap';
import { Actions } from '../state/Actions';

export class DisconnectedRequestJokeComponent extends React.Component {

    state: {
        input: string
    };

    constructor() {
        super();
        (this: any)._onInputChange = this._onInputChange.bind(this);
        (this: any)._onSubmit = this._onSubmit.bind(this);

        // setup initial state
        this.state = {
            input: ''
        }
    }

    _onInputChange(e: any) {
        this.setState({
            input: e.target.value
        });
    }

    _onSubmit() {
        const parsedInt = parseInt(this.state.input, 10);
        if (!isNaN(parsedInt)) {
            this.props.requestJoke(parsedInt);
            this.setState({
                input: ''
            });
        }
    }

    render() {
        return (
            <Panel>
                <FormGroup>
                    <ControlLabel>Request more Jokes!</ControlLabel>
                    <FormControl type="number"
                                 className="form-control-test"
                                 value={this.state.input}
                                 onChange={this._onInputChange}
                                 placeholder="Enter id"/>
                </FormGroup>

                <Button bsStyle="primary" onClick={this._onSubmit} className="test-submit-button">
                    Request joke with this id
                </Button>
            </Panel>
        );
    }
}

DisconnectedRequestJokeComponent.propTypes = {
    // injected through redux
    requestJoke: React.PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch: any) => {
    return {
        requestJoke: (num) => dispatch(Actions.requestJoke(num))
    };
};

export const RequestJokeComponent = connect(null, mapDispatchToProps)(DisconnectedRequestJokeComponent);
