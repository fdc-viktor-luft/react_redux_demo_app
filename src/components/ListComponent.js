// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { ListGroup, ListGroupItem, Button } from 'react-bootstrap';

import type { StoreType } from '../state/StoreProvider';
import { Jokes } from "../business/Jokes";
import { Actions } from "../state/Actions";

export class DisconnectedListComponent extends Component {

    constructor() {
        super();
    }

    _deleteJoke(position: number) {
        this.props.deleteJoke(position);
    }

    render() {
        return (
            <ListGroup fill>
                {
                    this.props.jokes.getJokes().map((value, index) => {
                        return (
                            <ListGroupItem key={'joke-' + index} className={'clearfix list-group-item-test-' + index}>
                                {value}
                                <Button bsSize="sm" className="pull-right" onClick={() => this._deleteJoke(index)}>Delete Joke!</Button>
                            </ListGroupItem>
                        )
                    }).toArray()
                }
            </ListGroup>
        );
    }
}

DisconnectedListComponent.propTypes = {
    // injected through redux
    jokes: React.PropTypes.instanceOf(Jokes).isRequired,

    deleteJoke: React.PropTypes.func.isRequired
};

const mapStateToProps = (state: StoreType) => {
    return {
        jokes: state.jokes
    };
};

const mapDispatchToProps = (dispatch: any) => {
    return {
        deleteJoke: (position) => dispatch(Actions.deleteJoke(position))

    };
};

export const ListComponent = connect(mapStateToProps, mapDispatchToProps)(DisconnectedListComponent);

