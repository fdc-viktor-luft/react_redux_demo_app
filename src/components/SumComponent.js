// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Well } from 'react-bootstrap';

import type { StoreType } from '../state/StoreProvider';
import { Jokes } from "../business/Jokes";

export class DisconnectedSumComponent extends Component {
    render() {
        return (
            <Well className="test-class-sum">{'Sum of all characters: ' + this.props.jokes.getJokes().reduce((red, elem) => red + elem.length, 0) }</Well>
        );
    }
}

DisconnectedSumComponent.propTypes = {
    // injected through redux
    jokes: React.PropTypes.instanceOf(Jokes).isRequired
};

export const SumComponent = connect(
    (state: StoreType) => {
        return {
            jokes: state.jokes
        };
    }
)(DisconnectedSumComponent);
