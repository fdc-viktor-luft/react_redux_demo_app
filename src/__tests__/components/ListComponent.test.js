import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Button } from 'react-bootstrap';

import { DisconnectedListComponent } from '../../components/ListComponent';
import { Jokes } from "../../business/Jokes";

describe('ListComponent', () => {
    it('[snapshot] it renders correctly', () => {
        const app = shallow(<DisconnectedListComponent jokes={new Jokes()} deleteJoke={() => {}}/>);
        expect(toJson(app)).toMatchSnapshot();
    });

    it('renders a list group item per joke', () => {
        const jokes = new Jokes().addNewJoke('this is a totally awesome joke!');
        const app = shallow(<DisconnectedListComponent jokes={jokes} deleteJoke={() => {}}/>);

        const element = app.find('.list-group-item-test-0');
        expect(element.contains('this is a totally awesome joke!')).toBe(true);
    });

    it('calls correct handler for the delete button', () => {
        const jokes = new Jokes().addNewJoke('this is a totally awesome joke!')
                                 .addNewJoke('this is a pretty lame joke...');

        const spy = jest.fn();
        const app = shallow(<DisconnectedListComponent jokes={jokes} deleteJoke={spy}/>);

        const btn = app.find('.list-group-item-test-1').find(Button);
        btn.simulate('click');

        expect(spy).toHaveBeenCalledWith(1);
    });
});
