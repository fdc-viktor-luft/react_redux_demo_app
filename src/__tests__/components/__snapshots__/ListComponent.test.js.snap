exports[`ListComponent [snapshot] it renders correctly 1`] = `
<ListGroup
  bsClass="list-group"
  fill={true} />
`;
